aptos-move/aptos-vm/src/aptos_vm.rs

```diff
@@ -1653,326 +1856,25 @@ impl VMValidator for AptosVM {
-                                    )?;
-
-                                self.0.success_transaction_cleanup(
-                                    respawned_session,
-                                    &gas_meter,
-                                    &txn_data,
-                                    log_context,
-                                    &storage_gas_params.change_set_configs,
-                                )
-                            })
-                        },
-                    }
-                } else {
-                    Err(VMStatus::error(StatusCode::MISSING_DATA, None))
-                }
-            },
-
-            // Deprecated. Will be removed in the future.
-            TransactionPayload::ModuleBundle(m) => self.0.execute_modules(
-                resolver,
-                session,
-                &mut gas_meter,
-                &txn_data,
-                m,
-                log_context,
-                &mut new_published_modules_loaded,
-                &storage_gas_params.change_set_configs,
-            ),
-        };
-
-        match result {
-            Ok(output) => output,
-            Err(err) => {
-                // Invalidate the loader cache in case there was a new module loaded from a module
-                // publish request that failed.
-                // This ensures the loader cache is flushed later to align storage with the cache.
-                // None of the modules in the bundle will be committed to storage,
-                // but some of them may have ended up in the cache.
-                if new_published_modules_loaded {
-                    self.0 .0.mark_loader_cache_as_invalid();
-                };
-                let txn_status = TransactionStatus::from_vm_status(
-                    err.clone(),
-                    self.0
-                         .0
-                        .get_features()
-                        .is_enabled(FeatureFlag::CHARGE_INVARIANT_VIOLATION),
-                );
-                if txn_status.is_discarded() {
-                    discard_error_vm_status(err)
-                } else {
-                    let (vm_status, output) = self.0.failed_transaction_cleanup_and_keep_vm_status(
-                        err,
-                        &gas_meter,
-                        &txn_data,
-                        resolver,
-                        log_context,
-                        &storage_gas_params.change_set_configs,
-                    );
-                    (vm_status, output)
-                }
-            },
-        }
+        transaction: SignedTransaction,
+        state_view: &(impl StateView + Sync),
+    ) -> TransactionOutput {
+        // The caller must ensure that the signature is not invalid, as otherwise
+        // this can be an attack.
+        let txn = SignatureVerifiedTransaction::valid_for_simulation(transaction)
+            .expect("Simulated transaction should not have a valid signature");
```

Comment:
I used expect because API has a check for that before calling simulation, and does the proper error handling. Do you think it is better 2 check both times?

Reply:
I see - not a major concern then. I think the best is if we can fold that check here (instead of at the caller). Otherwise, I'm fine with as is.

Reply:
The problem is that on API side we need to check anyway to differentiate between valid signature error and other errors?